### ISSUE TITLE
in_shardings and out_shardings cannot be passed to eqx.filter_jit

### ISSUE THREAD

**Issue by mjo22** [patrick-kidger/equinox#1077]
Hi, I’ve been discussing in #1076 and #1067 some thoughts for supporting automatic parallelization via `eqx.filter_jit` with something like the following interface:

```python
s, r = sharding, replicated

@eqx.filter_jit(donate=“all”, in_shardings=(eqx.if_array(r), eqx.if_array(s)), out_shardings=(eqx.if_array(r),)
def train_step(model_opt_state, x_y):
    model, opt_state = model_opt_state
    x, y = x_y
    …
```

I wanted to open a separate issue because I noticed that `eqx.filter_pmap` documentation says “using [equinox.filter_jit](https://docs.kidger.site/equinox/api/transformations/#equinox.filter_jit) with sharded inputs is now recommended over filter_pmap.”

I tried this and got some strange errors, and from looking at the code I realized this can’t actually be done in practice right now because the order of arguments passed ultimately to `jax.jit` is shuffled because of how the donate interface works. To achieve this recommendation, I think an explicit update is required. And further, I think it would be great to add the niceness of `eqx.filter_vmap` `AxisSpec`s.

**Comment by patrick-kidger** [comment]
Picking up the conversation from #1076: this was intentional in the design of `eqx.filter_jit` - as `{in,out}_shardings` were considered soft-deprecated, it didn't matter that there was no way to pass them.

If that's changed then we may need to adjust this.

**Comment by michael-0brien** [comment]
Closing this as per discussion in #1076
